Signal/src/ViewControllers/ContactViewController.swift

```diff
@@ -0,0 +1,686 @@
+
+        UIUtil.applySignalAppearence()
+
+        self.becomeFirstResponder()
+    }
+
+    private var scrollView: UIScrollView
+
+    override func loadView() {
+        super.loadView()
+
+        self.view.addSubview(scrollView)
+        scrollView.layoutMargins = .zero
+        scrollView.autoPinWidthToSuperview()
+        scrollView.autoPin(toTopLayoutGuideOf: self, withInset: 0)
+        scrollView.autoPin(toBottomLayoutGuideOf: self, withInset: 0)
+
+        self.view.backgroundColor = UIColor.white
+
+        updateContent()
+
+        hasLoadedView = true
+    }
+
+    private func tryToDetermineMode() {
+        SwiftAssertIsOnMainThread(#function)
+
+        guard phoneNumbersForContact().count > 0 else {
+            viewMode = .noPhoneNumber
+            return
+        }
+        if systemContactsWithSignalAccountsForContact().count > 0 {
+            viewMode = .systemContactWithSignal
+            return
+        }
+        if systemContactsForContact().count > 0 {
+            viewMode = .systemContactWithoutSignal
+            return
+        }
+
+        viewMode = .nonSystemContact
+    }
+
+    private func systemContactsWithSignalAccountsForContact() -> [String] {
+        SwiftAssertIsOnMainThread(#function)
+
+        return phoneNumbersForContact().filter({ (phoneNumber) -> Bool in
+            return contactsManager.hasSignalAccount(forRecipientId: phoneNumber)
+        })
+    }
+
+    private func systemContactsForContact() -> [String] {
+        SwiftAssertIsOnMainThread(#function)
+
+        return phoneNumbersForContact().filter({ (phoneNumber) -> Bool in
+            return contactsManager.allContactsMap[phoneNumber] != nil
+        })
+    }
+
+    private func phoneNumbersForContact() -> [String] {
+        SwiftAssertIsOnMainThread(#function)
+
+        var result = [String]()
+        if let phoneNumbers = contact.phoneNumbers {
+            for phoneNumber in phoneNumbers {
+                result.append(phoneNumber.phoneNumber)
+            }
+        }
+        return result
+    }
+
+    private func tryToDetermineModeRetry() {
+        // Try again after a minute.
+        DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + 60.0) { [weak self] in
+            guard let strongSelf = self else { return }
+            strongSelf.tryToDetermineMode()
+        }
+    }
+
+    private func updateContent() {
+        SwiftAssertIsOnMainThread(#function)
+
+        let rootView = self.scrollView
+
+        for subview in rootView.subviews {
+            subview.removeFromSuperview()
+        }
+
+        // TODO: The design calls for no navigation bar, just a back button.
+        let topView = UIView.container()
+        topView.backgroundColor = UIColor(rgbHex: 0xefeff4)
+        topView.preservesSuperviewLayoutMargins = true
+        rootView.addSubview(topView)
+        topView.autoPinEdge(toSuperviewEdge: .top)
+        topView.autoPinEdge(.left, to: .left, of: self.view, withOffset: 0)
```

Comment:
Except for very rare circumstances, prefer .leading/.trailing instead of .left/.right.

I'd say any usage of left/right probably deserves a justifying comment.